Multi client

zthatch opened the pull request:
Added a client wrapper class to allow for clients across different databases. Tested manually rather than implementing a new conftest fixture. Next step is potentially to implement an additional conftest fixture and to create an interface metaclass for the clients to inherit from, requiring that client classes have specific methods with specific arguments.

zthatch commented:
Take a close look at what I've done to try and ensure backwards compatibility with old regolithrc.json files. The big change is that I've put the backend key inside of the databases dictionaries. I try to account for old configurations in my logic.

sbillinge commented:
@scopatz @CJ-Wright I know you are both very busy but this is a change to the regolith infrastructure, so I would love it if you could take a look.   

The goal:
to be able to mix backends.  I want to start migrating db's from filesystem to mongo clients, but I would rather do it one db at a time.  to do that we have to move `rc.host` inside the `database` document.  I asked Zach to do it in a way that doesn't break backwards compatibility.  He thinks he has succeeded on all counts.

If you think it looks ok, we will go ahead and write a test fixture for it and a news item.  I can also start to test it on a branch, but honestly, the best way to test it is to merge it (if it doesn't break anything) and try it out on actual databases.

Also, I think we should think about a new release of regolith and clear out that news directory.  None of the helper infrastructure has been released.  Would you be ok with this?

Happy new year, I hope you are both well!

sbillinge reviewed (commented):
few small comments.  See inline.

sbillinge commented:
@zthatch I forgot but we definitely need a news item for this.

sbillinge commented:
@scopatz @CJ-Wright I want to merge this, but I won't for a few days in case you want to look at it.

zthatch commented on regolith/mongoclient.py:
```diff
@@ -314,6 +318,8 @@ def insert_one(self, dbname, collname, doc):
         """Inserts many documents into a database/collection."""
         coll = self.client[dbname][collname]
+        for doc in docs:
+            doc['_id'].replace('.', '')
```
The reason I know that this bit of code is necessary is because it is used in the conftest.py code when creating the mongo test fixture. There actually are _ids in the schema exemplars with dots, but they get removed in the creation of the test fixture.

If we want to test them, I'll just take that out.

Whenever I wrote the test initially I must have been more focused on making the test pass than making sure the code worked. Big wupsie daisy. 

Not sure that we need a decorator since this is specific to mongo backend collection entry, so it shouldn't actually be repeated that often. Let me know why you are thinking decorator. Maybe at morning meeting.

zthatch commented on tests/conftest.py:
```diff
@@ -44,6 +44,8 @@ def make_db():
                         "public": True,
                         "path": "db",
                         "local": True,
+                        #TODO uncomment this out after proving JSON w/o key worrk
```
deleted

zthatch commented on regolith/client_manager.py:
```diff
@@ -0,0 +1,131 @@
+        for database in databases:
+            if "backend" not in database:
+                # For backwards compatabity, since most people using an FS backend won't have the backend keyword in rc
+                database["backend"] = 'filesystem'
```
added info block

zthatch commented:
@sbillinge I addressed your comments and added a test to make sure the new functionality is actually being tested. 

Unfortunately, the code coverage doesn't see that the client manager code is being tested. Maybe it wants there to be tests for all of those specific functions, even though it is wrapping code used in tests that already exists and therefore is being proven out by virtue of the fact that nothing has broken. This is just a guess.

I'm not as familiar with code coverage testing as I should be, but I did see some "#no pragma" code or something like that in SHED, so maybe I have to take a page out of that book.

CJ-Wright commented:
FYI I don't think I'll be able to look this over in the near future.

sbillinge commented:
- [] make a develop branch based on latest master
- [] change this PR to be into that branch
- [] zach set up some local collection all the way to atlas and make docs for how to do it
- [] set up a call for Zach to help Simon set up things up locally this way.
- [] let simon use it for his todo list, so it will get a really great workout
- [] fix any issues
- [] merge into master

sbillinge commented:
notes:
 - () make a develop branch based on latest master
[] change this PR to be into that branch
[] zach set up some local collection all the way to atlas and make docs for how to do it

sbillinge commented:
@zthatch please let me know when you are ready for me to merge.  This is passing tests except for some diff codecov target which we can overlook.

zthatch commented:
> @zthatch please let me know when you are ready for me to merge. This is passing tests except for some diff codecov target which we can overlook.

@sbillinge After tests it is ready to merge. I just rebased with master so your datetime fix should be implemented. Everything is working. 

Next step is definitely validation, but I am going to shift for the next two days back to getxgui, and then shift back when we meet on Thursday.

sbillinge commented:
just a quick question....will it cause problems later if we rebase your branch on to updated master, then want to merge back into master later?  Will it just ignore the repeated commits?   I never did this workflow before with a develop branch that can diverge from master......any ideas?  

We could also merge master into develop and then merge your branch without the rebase.  Could you look into it quickly?  Thnaks
